webrtc/stun/stun_connection.nim

```diff
@@ -0,0 +1,249 @@
+  StunPasswordProvider* = proc(username: seq[byte]): seq[byte] {.raises: [], gcsafe.}
+
+  StunConn* = ref object
+    conn*: UdpConn # Underlying UDP connexion
+    laddr*: TransportAddress # Local address
+    raddr*: TransportAddress # Remote address
+    dataRecv*: AsyncQueue[seq[byte]] # data received which will be read by DTLS
+    stunMsgs*: AsyncQueue[seq[byte]] # stun messages received and to be
+                                     # processed by the stun message handler
+    handlesFut*: Future[void] # Stun Message handler
+    closeEvent: AsyncEvent
+    closed*: bool
+
+    # Is ice-controlling and iceTiebreaker, not fully implemented yet.
+    iceControlling: bool
+    iceTiebreaker: uint32
+
+    # Specified by the user
+    usernameProvider: StunUsernameProvider
+    usernameChecker: StunUsernameChecker
+    passwordProvider: StunPasswordProvider
+
+    rng: ref HmacDrbgContext
+
+# - Create Binding Messages (Request / Response / Error)
+
+proc getBindingResponse*(self: StunConn, msg: StunMessage): StunMessage =
+  ## Takes an encoded Stun Message and the local address. Returns
+  ## an encoded Binding Response if the received message is a
+  ## Binding request.
+  ##
+  result = StunMessage(msgType: StunBindingResponse,
+                        transactionId: msg.transactionId)
+  result.attributes.add(XorMappedAddress.encode(self.raddr, msg.transactionId))
+
+proc calculatePriority(self: StunConn): uint32 =
+  # https://datatracker.ietf.org/doc/html/rfc8445#section-5.1.2.1
+  # Calculate Ice priority. At the moment, we assume we're a publicly available server.
+  let typePreference = 126'u32
+  let localPreference = 65535'u32
+  let componentID = 1'u32
+  return (1 shl 24) * typePreference + (1 shl 8) * localPreference + (256 - componentID)
+
+proc getBindingRequest*(self: StunConn): StunMessage =
+  ## Creates an encoded Binding Request
+  ##
+  result = StunMessage(msgType: StunBindingRequest)
+  self.rng[].generate(result.transactionId)
+
+  let username = self.usernameProvider()
+  if username != "":
+    result.attributes.add(UsernameAttribute.encode(username))
+
+  if self.iceControlling:
+    result.attributes.add(IceControlling.encode(self.iceTiebreaker))
+  else:
+    result.attributes.add(IceControlled.encode(self.iceTiebreaker))
+  result.attributes.add(Priority.encode(self.calculatePriority()))
+
+proc checkForError*(self: StunConn, msg: StunMessage): Option[StunMessage] =
+  # Check for error from a BindingRequest message.
+  # Returns an option with some BindingErrorResponse if there is an error.
+  # Returns none otherwise.
+  # https://datatracker.ietf.org/doc/html/rfc5389#section-10.1.2
+  var res = StunMessage(msgType: StunBindingErrorResponse,
+                        transactionId: msg.transactionId)
+  if msg.getAttribute(MessageIntegrity).isNone() or
+     msg.getAttribute(UsernameAttribute).isNone():
+    res.attributes.add(ErrorCode.encode(ECBadRequest))
+    return some(res)
```

Comment:
What do you think about returning `StunMessage(msgType: StunBindingErrorResponse, transactionId: msg.transactionId, attributes: @[ErrorCode.encode(ECBadRequest)])` instead of creating a msg every single time this proc is called, even if there's no error?